### ISSUE TITLE
bottleneck to uploading from users with closed ports & autoexpanding downloading lists

### ISSUE THREAD

**Issue by xristos10** [nicotine-plus/nicotine-plus#3033]
Description:

Issue 1 - uploading bottleneck
Let's say I have 500KiB/s uploading bandwidth and there are 2 users that want to download from me: user.no1 user.no2
User.no1 for some reason (usually a blocked port) doesn't use all the available bandwidth that I offer, he downloads with 5KiB/s.
Then User.no2 starts a download too but he gets 250KiB/s max, the rest of the uploader's available bandwidth seems to be reserved by User.no1.
This is a very common situation.

Issue 2 - autoexpanding downloading lists
In the uploads tab, set on 'Group by user' I click the 'collapse all' arrow button, but if a downloader has more than 1 file in queue the lists keep auto expanding all the time, no matter if I close it again.
This started a couple versions ago, I currently use Nicotine+ 3.3.4, I uninstalled, switched from installation to portable but no luck.

Sorry in advance if these have been mentioned before, keep up the great job you do, especially in the recent months.

**Comment by mathiascode** [comment]
We could do a better job distributing the bandwidth among users. Right now the maximum speed per user is just the total speed divided by the number of active transfers.

The second issue is https://github.com/nicotine-plus/nicotine-plus/issues/2969, which will be fixed in 3.3.5.

**Comment by xristos10** [comment]
Thanks for the prompt answer.

The utilization of the uploading bandwidth is a core thing cause closed ports are a frequent problem since the days of legacy soulseek (now even worse with failed usage of vpns), and it's not just a problem of the dude with the closed port as it affects the others too. In the end this is all about transferring  : )

All that can be done for now is to remove the misconfigured user manually, until he connects again later on.

**Comment by mathiascode** [comment]
A closed port shouldn't affect transfer speeds. If a user's port is closed, they just end up connecting to your port instead, which establishes a normal connection. Some users most likely just have slow download speeds.

**Comment by slook** [comment]
@xristos10 Although this issue isn't resolved yet, improvements to the calculation of Speed measurements are in the latest development build of Nicotine+ 3.3.5.dev2 please test it out.

**Comment by xristos10** [comment]
Of course, i ll get back with a short feedback in a week or less. Thanks.
